release v2 use aws-sdk-v2

BobDu opened the pull request.

vieux commented:
👍 would be a good idea to have this merged.

vieux commented:
ping @a8m, I see you're pretty active on GH. Are you interested in merging this?

BobDu commented:
@a8m ping.

With AWS announcing the end of support for AWS SDK for Go v1 by July 31, 2025 (https://aws.amazon.com/blogs/developer/announcing-end-of-support-for-aws-sdk-for-go-v1-on-july-31-2025/), This upgrade seems to be more necessary now.

Could you take some time to review this PR and release a new version?

a8m commented on aggregator.go:
```diff
@@ -53,28 +54,28 @@ func (a *Aggregator) Put(data []byte, partitionKey string) {
 // that compatible with the KCL's deaggregation logic.
 //
 // If you interested to know more about it. see: aggregation-format.md
-func (a *Aggregator) Drain() (*k.PutRecordsRequestEntry, error) {
+func (a *Aggregator) Drain() (ktypes.PutRecordsRequestEntry, bool, error) {
 	if a.nbytes == 0 {
```
Why did you change the signature and not just changed the type?

a8m commented on .gitignore:
```diff
@@ -1 +1,2 @@
 draft
+.idea
```
Use global .gitignore instead. This file should be removed from here completely

a8m commented on go.mod:
```diff
@@ -1,12 +1,17 @@
 module github.com/a8m/kinesis-producer
 
+go 1.16
```
upgrade to 1.23

a8m commented on producer.go:
```diff
@@ -215,26 +219,26 @@ func (p *Producer) loop() {
 	if needToDrain {
 		p.Lock()
-		record, err := p.aggregator.Drain()
+		record, has, err := p.aggregator.Drain()
```
Let's not mix logic change with SDK upgrade. Also, if the previous code uses pointers, keep using pointers.

a8m commented on producer_test.go:
```diff
@@ -20,7 +22,7 @@ type clientMock struct {
 }
 
-func (c *clientMock) PutRecords(input *k.PutRecordsInput) (*k.PutRecordsOutput, error) {
+func (c *clientMock) PutRecords(ctx context.Context, input *k.PutRecordsInput, optFns ...func(*k.Options)) (*k.PutRecordsOutput, error) {
```
```suggestion
func (c *clientMock) PutRecords(_ context.Context, input *k.PutRecordsInput, _ ...func(*k.Options)) (*k.PutRecordsOutput, error) {
```

a8m commented on producer.go:
```diff
@@ -243,8 +247,8 @@ func (p *Producer) flush(records []*kinesis.PutRecordsRequestEntry, reason strin
 		p.Logger.Info("flushing records", LogValue{"reason", reason}, LogValue{"records", len(records)})
-		out, err := p.Client.PutRecords(&kinesis.PutRecordsInput{
-			StreamName: &p.StreamName,
+		out, err := p.Client.PutRecords(context.TODO(), &kinesis.PutRecordsInput{
```
```suggestion
		out, err := p.Client.PutRecords(context.Background(), &kinesis.PutRecordsInput{
```

a8m reviewed (commented):
Thanks for the contribution, @BobDu.

Sorry for my delayed review. I wrote this package nine years ago when I built my product on top of Kinesis, but I currently do not use this service, and therefore, this package is not actively maintained.

BobDu commented:
https://github.com/aws/aws-sdk-go/blob/825250a3f2f45ff9322c4a9ae2dd96e5bdb93ea4/service/kinesis/api.go#L8237

https://github.com/aws/aws-sdk-go-v2/blob/65023f312c12ce2707836e552ab55bb4f03f3345/service/kinesis/api_op_PutRecords.go#L92

Due to the modification of the `Records` field type in `PutRecordsInput` by `aws-go-sdk-v2`
`Records []*PutRecordsRequestEntry` => `Records []types.PutRecordsRequestEntry`

So, I have also modified some function signatures to no longer use pointer slices.
However, for non-slice type parameters, I have reverted to the previous behavior.

BobDu commented:
Additionally, I have updated the version of the gRPC library and other dependencies.

All comment are resolved. @a8m Please review it again.

BobDu commented:
Additionally, I have added support for the kinesis `StreamARN`

a8m commented:
> Additionally, I have added support for the kinesis `StreamARN`

Please, let’s not mix changes in the same PR. I want to approve and merge the first one.

BobDu commented:
> Please, let’s not mix changes in the same PR. I want to approve and merge the first one.

Sorry, I revert this change.
If you think it is necessary, I can revert update protobuf dependencies commit same.

And I will create other PR.  @a8m

a8m reviewed (approved):
💯 

Thanks for the contribution, @BobDu
